model/historical_test_data.go

```diff
@@ -258,3 +269,484 @@ func (i *HistoricalTestDataInfo) validate() error {
+type HTDGroupBy string
+
+const (
+	HTDGroupByTest    HTDGroupBy = "test"
+	HTDGroupByTask    HTDGroupBy = "task"
+	HTDGroupByVariant HTDGroupBy = "variant"
+)
+
+func (gb HTDGroupBy) validate() error {
+	switch gb {
+	case HTDGroupByVariant:
+	case HTDGroupByTask:
+	case HTDGroupByTest:
+	default:
+		return errors.Errorf("invalid HTDGroupBy value: %s", gb)
+	}
+
+	return nil
+}
+
+// HTDSort represents the sorting options for historical test data.
+type HTDSort string
+
+const (
+	HTDSortEarliestFirst HTDSort = "earliest"
+	HTDSortLatestFirst   HTDSort = "latest"
+)
+
+func (s HTDSort) validate() error {
+	switch s {
+	case HTDSortEarliestFirst:
+	case HTDSortLatestFirst:
+	default:
+		return errors.Errorf("invalid HTDSort value: %s", s)
+	}
+
+	return nil
+}
+
+// HTDStartAt represents parameters that allow a search query to resume at a
+// specific point. Used for pagination.
+type HTDStartAt struct {
+	Date    time.Time
+	Variant string
+	Task    string
+	Test    string
+}
+
+func (s *HTDStartAt) validate(groupBy HTDGroupBy) error {
+	catcher := grip.NewBasicCatcher()
+	if s == nil {
+		catcher.New("StartAt should not be nil")
+	}
+	if !s.Date.Equal(utility.GetUTCDay(s.Date)) {
+		catcher.New("invalid StartAt Date value")
+	}
+	if len(s.Test) == 0 {
+		catcher.New("missing StartAt Test value")
+	}
+	switch groupBy {
+	case HTDGroupByVariant:
+		if len(s.Variant) == 0 {
+			catcher.New("missing StartAt Variant value")
+		}
+		fallthrough
+	case HTDGroupByTask:
+		if len(s.Task) == 0 {
+			catcher.New("missing StartAt Task value")
+		}
+	}
+
+	return catcher.Resolve()
+}
+
+// HistoricalTestDataFilter represents search and aggregation parameters when
+// querying the historical test data.
+type HistoricalTestDataFilter struct {
+	Project    string
+	Requesters []string
+	AfterDate  time.Time
+	BeforeDate time.Time
+
+	Tests    []string
+	Tasks    []string
+	Variants []string
+
+	GroupNumDays int
+	GroupBy      HTDGroupBy
+	StartAt      *HTDStartAt
+	Limit        int
+	Sort         HTDSort
+}
+
+func (f *HistoricalTestDataFilter) validate() error {
+	catcher := grip.NewBasicCatcher()
+	if f == nil {
+		catcher.New("historical test data filter should not be nil")
+	}
+
+	if len(f.Requesters) == 0 {
+		catcher.New("missing Requesters values")
+	}
+	if f.GroupNumDays <= 0 {
+		catcher.New("invalid GroupNumDays value")
+	}
+	if f.Limit > htdMaxQueryLimit || f.Limit <= 0 {
+		catcher.New("invalid Limit value")
+	}
+	if f.StartAt != nil {
+		catcher.Add(f.StartAt.validate(f.GroupBy))
+	}
+	if len(f.Tests) == 0 && len(f.Tasks) == 0 {
+		catcher.New("missing Tests or Tasks values")
+	}
+
+	catcher.Add(f.Sort.validate())
+	catcher.Add(f.GroupBy.validate())
+	catcher.Add(f.validateDates())
+
+	return catcher.Resolve()
+}
+
+func (f *HistoricalTestDataFilter) validateDates() error {
+	catcher := grip.NewBasicCatcher()
+	if !f.AfterDate.Equal(utility.GetUTCDay(f.AfterDate)) {
+		catcher.New("invalid AfterDate value")
```

Comment:
Can we make it more obvious that the AfterDate/BeforeDate is only valid if it's equal to midnight UTC of a day?

Reply:
That is done in user facing documentation: https://github.com/evergreen-ci/evergreen/wiki/REST-V2-Usage#teststats